beacon-chain/light-client-network.md

```diff
@@ -0,0 +1,179 @@
+a node is "interested" in.
+
+We define the following function to determine whether node in the network should be interested in a piece of content:
+
+```
+interested(node, content) = distance(node.id, content.id) <= node.radius
+```
+
+A node is expected to maintain `radius` information for each node in its local node table. A node's `radius` value may fluctuate as the contents of its local key-value store change.
+
+A node should track their own radius value and provide this value in all Ping or Pong messages it sends to other nodes.
+
+### Data Types
+
+The beacon chain light client DHT stores the following data items:
+
+* LightClientBootstrap
+* LightClientUpdate
+
+The following data objects are ephemeral and we store only the latest values:
+
+* LightClientFinalityUpdate
+* LightClientOptimisticUpdate
+
+#### Constants
+
+We define the following constants which are used in the various data type definitions:
+
+```py
+# Maximum number of `LightClientUpdate` instances in a single request
+MAX_REQUEST_LIGHT_CLIENT_UPDATES = 2**7  # = 128
+```
+
+#### LightClientBootstrap
+
+```
+light_client_bootstrap_key = Container(block_hash: Bytes32)
+selector                   = 0x00
+
+content                    = SSZ.serialize(LightlientBootstrap)
+content_key                = selector + SSZ.serialize(light_client_bootstrap_key)
+```
+
+#### LightClientUpdatesByRange
+
+```
+light_client_update_keys   = Container(start_period: uint64, count: uint64)
+selector                   = 0x01
+
+content                    = SSZList(LightClientUpdate, max_lenght=MAX_REQUEST_LIGHT_CLIENT_UPDATES)
+content_key                = selector + SSZ.serialize(light_client_update_keys)
+```
+
+#### LightClientFinalityUpdate
+
+```
+light_client_finality_update_key  = Container(None)
+selector                          = 0x02
+
+content                           = SSZ.serialize(light_client_finality_update)
+content_key                       = selector + SSZ.serialize(light_client_finality_update_key)
+```
+
+> A `None` in the content key is equivalent to the request for the latest
+LightClientFinalityUpdate that the requested node has available.
+
+#### LightClientOptimisticUpdate
+
+```
+light_client_optimistic_update_key   = Container(None)
+selector                             = 0x03
+
+content                              = SSZ.serialize(light_client_optimistic_update)
+content_key                          = selector + SSZ.serialize(light_client_optimistic_update_key)
+```
+
+> A `None` in the content key is equivalent to the request for the latest
+LightClientOptimisticUpdate that the requested node has available.
+
+### Algorithms
+
+#### Portal Gossip
+
+TODO
+
+#### Validation
+
+Validating `LightClientFinalityUpdate` and `LightClientOptimisticUpdate` follows the gossip domain(gossipsub) [consensus specs](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#the-gossip-domain-gossipsub).
```

Comment:
There is an important difference compared to other portal network data that is specific to light client data, that there can be multiple valid objects with conflicting content keys due to the way how aggregate signatures work.

The [`is_better_update`](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#is_better_update) function should be used when encountering conflicting data to resolve conflicts. This ensures that if conflicting data is encountered, the one with higher participation in the signature takes precedence.

Reply:
Note that this also affects propagation, i.e., when encountering a better update than the local one, the better one should be propagated among peers. Maybe the update quality needs to be encoded into the message id for this purpose.